### PR TITLE
Invariant cleanup, util directory, files j-r

### PR DESCRIPTION
**Pull request by sonodtt**
Cleanup of util directory, files j-r (json_irep.cpp withheld awaiting a review of cbmc exceptions)

1. Cleanup of asserts - replaced with invariants.

2. Removed all unstructured `throw(...)` statements, replacing with `INVARIANT`, `PRECONDITION`, where appropriate.

### CODE REVIEW

File: src/util/mp_arith.cpp [hunk 1]
```diff
@@ -280,8 +280,7 @@ mp_integer arith_right_shift(
   PRECONDITION(a.is_long() && b.is_ulong());
   llong_t number=a.to_long();
   ullong_t shift=b.to_ulong();
-  if(shift>true_size)
-    throw "shift value out of range";
+  DATA_INVARIANT(shift<=true_size, "shift value must be in range");
```
Comment (hannes-steffenhagen-diffblue) [inline src/util/mp_arith.cpp]: `PRECONDITION`

**Review by hannes-steffenhagen-diffblue — CHANGES_REQUESTED** [review]
One invariant I'd change, also please run `git-clang-format`

**Comment by sonodtt** [comment]
@hannes-steffenhagen-diffblue  (throws ashes on head ) - I did not push local changes. :(
Done.

File: src/util/mp_arith.cpp [hunk 2]
```diff
@@ -280,8 +280,7 @@ mp_integer arith_right_shift(
   PRECONDITION(a.is_long() && b.is_ulong());
   llong_t number=a.to_long();
   ullong_t shift=b.to_ulong();
-  if(shift>true_size)
-    throw "shift value out of range";
+  DATA_INVARIANT(shift <= true_size, "shift value must be in range");
```
Comment (hannes-steffenhagen-diffblue) [inline src/util/mp_arith.cpp]: This should still be a `PRECONDITION`

File: src/util/reference_counting.h [hunk 3]
```diff
@@ -121,9 +123,9 @@ void reference_counting<T>::remove_ref(dt *old_d)
   if(old_d==nullptr)
     return;
 
-  assert(old_d->ref_count!=0);
+  PRECONDITION(old_d->ref_count != 0);
 
-  #ifdef REFERENCE_COUNTING_DEBUG
```
Comment (hannes-steffenhagen-diffblue) [inline src/util/reference_counting.h]: This seems like an odd change, was this done by clang-format?

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
OK, minor :pick: aside

File: src/util/reference_counting.h (hunk 3 above)
Comment (sonodtt) [reply src/util/reference_counting.h]: I have undone those indentations and it did not happen again upon a clang-format. (!?!)

File: src/util/mp_arith.cpp (hunk 2 above)
Comment (sonodtt) [reply src/util/mp_arith.cpp]: Yep. It is now that I pushed the local changes. (Doh!)

File: src/util/namespace.h [hunk 4]
```diff
@@ -33,8 +34,11 @@ class namespace_baset
   const symbolt &lookup(const irep_idt &name) const
   {
     const symbolt *symbol;
-    if(lookup(name, symbol))
-      throw "identifier "+id2string(name)+" not found";
+    bool not_found = lookup(name, symbol);
+    INVARIANT(
+      !not_found,
+      "we are assuming that a name exists in the namespace "
+      "when this function is called");
```
Comment (tautschnig) [inline src/util/namespace.h]: Please include/keep `id2string(name)` in the error message.

File: src/util/rational.cpp [hunk 5]
```diff
@@ -10,6 +10,7 @@ Author: Daniel Kroening, kroening@kroening.com
 /// Rational Numbers
 
 #include "rational.h"
+#include "invariant.h"
```
Comment (tautschnig) [inline src/util/rational.cpp]: Maintain lexicographic ordering of includes: move one line up.

File: src/util/reference_counting.h [hunk 6]
```diff
@@ -14,6 +14,8 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <cassert>
```
Comment (tautschnig) [inline src/util/reference_counting.h]: Should now be removed.

**Review by tautschnig — APPROVED** [review]
Mostly looks good except as noted below.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 3a915bf0590f58e0cbc4f468d8ce9bc705cb26f4).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

File: src/util/rational.cpp (hunk 5 above)
Comment (sonodtt) [reply src/util/rational.cpp]: The first was the source file's own header.  My lack of a spacing line did not make this clear. 
Amended to:
```
#include "rational.h"

#include <algorithm>
#include <ostream>

#include "invariant.h"
```

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 341f99349eee592a0614b126fcfe8c4fcf734a4e).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 341f99349eee592a0614b126fcfe8c4fcf734a4e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/83828018

**Pull request merged**